### PR TITLE
Fixes #14410 - respond 503 when pending migration

### PR DESCRIPTION
**Pull request by iNecas**

### CODE REVIEW

File: app/controllers/concerns/foreman/controller/migration_checker.rb [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+  private
+
+  def check_pending_migrations
+    if Foreman::Controller::MigrationChecker.needs_migration?
+      raise Foreman::MaintenanceException, _("There are migrations pending in the system.")
+    end
+  end
+
+  def self.needs_migration?
+    return @needs_migration unless @needs_migration == nil
```
Comment (ares) [inline app/controllers/concerns/foreman/controller/migration_checker.rb]: `@needs_migration.nil?`

File: app/controllers/concerns/foreman/controller/migration_checker.rb [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+
+  included do
+    before_filter :check_pending_migrations
+  end
+
+  private
+
+  def check_pending_migrations
+    if Foreman::Controller::MigrationChecker.needs_migration?
+      raise Foreman::MaintenanceException, _("There are migrations pending in the system.")
```
Comment (dLobatog) [inline app/controllers/concerns/foreman/controller/migration_checker.rb:12]: Instructions on how to move forward would be helpful here: `foreman-rake db:migrate`

Comment (iNecas) [inline app/controllers/concerns/foreman/controller/migration_checker.rb:12]: The resolution might differ depending on how you perform the upgrades. I would rather keep the resolution described in http://projects.theforeman.org/projects/foreman/wiki/ErrorCodes

File: app/controllers/concerns/foreman/controller/migration_checker.rb (hunk 1 above)
Comment (iNecas) [inline app/controllers/concerns/foreman/controller/migration_checker.rb]: Right, fixed

**Comment by iNecas** [comment]
Intermittent issue [test]

**Comment by dLobatog** [comment]
I really don't think this is the place for maintenance mode - if we even need it I'd rather just call it from the installer instead of maintaining it ourselves. https://github.com/biola/turnout does a good job at that. I'll spare some time to send an email to foreman-dev and see what other devs think about it

**Comment by iNecas** [comment]
I'm not saying this is the ultimate solution to the maintenance mode and I would like to cover this properly with installer.

On the other hand, I prefer being defensive (if nothing else, not everyone is using installer in production) and I don't see a reason why there couldn't be more layers for making sure we can handle the request.

The only downside I've seen as argument so far was a concern about performance (an if statement on every request), which I do not think is really something that even benchmark would notice, especially when looking at https://github.com/biola/turnout/blob/355d8b39dc49958964bbf20d40ef195858ddc18b/lib/rack/turnout.rb#L17, where they at the end call File.exist? on every request. I don't think we need the new dependency for this.

Long story short:
1. I'm for us doing better job at installer level to handle the cases - this is not aiming to replace that but I don't see a reason why we can't have two checks
2. I don't think it's worth pulling in the dependency to address the issues we can address with a small patch like this: no rocket science here really
3. This patch is small enough to be a low hanging fruit until we make a proper changes in the installer process

**Comment by ares** [comment]
While we can discuss how to properly introduce maintenance mode on foreman-dev, this PR is orthogonal and only improves the error page. We can use this as one of entry points to maintenance mode when we have it properly implemented. Until then this is a nice way of handling the migration pending status. I suppose the exception name can be changed if `MaintenanceException` is something that indicates this introduces the maintanance mode.

FWIW, one new `if` per request is not something that should raise concerns I think. There are many other optimizations to be done before we get to point where this matters.

**Comment by dLobatog** [comment]
OK - I'll merge, let's be on the lookout for this so it doesn't grow too much :smile:

**Comment by dLobatog** [comment]
Merged as 2aa15bf1f40fff77a50bb9907fa993e067dd6346, thanks @iNecas!
